frontend/controller/service-worker.js

```diff
@@ -13,23 +16,120 @@ sbp('sbp/selectors/register', {
+
+      if (existingSubscription) {
+        // If there is an existing subscription, no need to create a new one.
+        // But make sure server knows the subscription details too.
+        pubsub.socket.send(createMessage(
+          NOTIFICATION_TYPE.PUSH_ACTION,
+          {
+            action: PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION,
+            payload: JSON.stringify(existingSubscription.toJSON())
+          }
+        ))
+        return
+      } else {
+        // Create a new push subscription
+
+        sbp('okTurtles.events/once', NOTIFICATION_TYPE.PUSH_ACTION, async ({ data }) => {
+          const PUBLIC_VAPID_KEY = data
+
+          // 1. Add a new subscription to pushManager using it.
+          const subscription = await registration.pushManager.subscribe({
+            userVisibleOnly: true,
+            applicationServerKey: urlBase64ToUint8Array(PUBLIC_VAPID_KEY)
+          })
+
+          // 2. Send the subscription details to the server. (server needs it to send the push notification)
+          pubsub.socket.send(createMessage(
+            NOTIFICATION_TYPE.PUSH_ACTION,
+            {
+              action: PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION,
+              payload: JSON.stringify(subscription.toJSON())
+            }
+          ))
+
+          // 3. Send the test notification to confirm it works as expected. (Just a demo for now, but can be removed after development)
+          const testNotification = {
+            title: 'Service worker installed.',
+            body: 'You can now receive various push notifications from the Group Income app!',
+            endpoint: subscription.endpoint
+          }
+
+          pubsub.socket.send(createMessage(
+            NOTIFICATION_TYPE.PUSH_ACTION,
+            {
+              action: PUSH_SERVER_ACTION_TYPE.SEND_PUSH_NOTIFICATION,
+              payload: JSON.stringify(testNotification)
+            }
+          ))
         })
-      }, 1000)
+
+        pubsub.socket.send(createMessage(
+          NOTIFICATION_TYPE.PUSH_ACTION,
+          { action: PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY }
+        ))
+      }
     } catch (e) {
-      // TODO: this
       console.error('error setting up service worker:', e)
     }
+  },
+  'service-worker/send-push': async function (payload) {
+    if (!('serviceWorker' in navigator)) { return }
+
+    const swRegistration = await navigator.serviceWorker.ready
+
+    if (!swRegistration) {
+      console.error('No service-worker registration found!')
+      return
+    }
+
+    const pushSubscription = await swRegistration.pushManager.getSubscription()
+    const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+
+    if (pushSubscription) {
+      pubsub.socket.send(createMessage(
+        NOTIFICATION_TYPE.PUSH_ACTION,
+        {
+          action: PUSH_SERVER_ACTION_TYPE.SEND_PUSH_NOTIFICATION,
+          payload: JSON.stringify({ ...payload, endpoint: pushSubscription.endpoint })
```

Comment:
I am wondering about this paragraph:

> The endpoint takes the form of a custom URL pointing to a push server, which can be used to send a push message to the particular service worker instance that subscribed to the push service. For this reason, it is a good idea to keep your endpoint a secret, so others do not hijack it and abuse the push functionality.
> 
> — https://developer.mozilla.org/en-US/docs/Web/API/PushSubscription/endpoint

I haven't had a chance to study this in depth. Wonder if you or @corrideat have any insights here as far as whether there are any security issues in how we're using this.

Reply:
I'm not entirely familiar with the implications of this URL (would need to see how this protocol works), although it seems, from a general POV, like URLs shouldn't be treated as secrets.

Until I review this, though, this is my preliminary assessment:

(1) Messages should be encrypted and signed. Providing this holds, the security risk from a 'public' URL should be limited to DoS at most.
(2) If the URL needs to be secret, we could derive it from a secret known to the client only (e.g., we could add a 'subscription secret' to the identity contract)